### ISSUE TITLE
[Bug]: Healthcheck not working as expected

### ISSUE THREAD

**Issue by EinfachHans** [coollabsio/coolify#2073]
### Description

After a longer email conversation with andras about health checks i want to track the last bugs around health-checks i have here. Thanks for all the help so far 👍🏼

Before i had this configuration in my dockerfile:
```
HEALTHCHECK --interval=10s --timeout=3s --start-period=30s \
 CMD curl -f http://localhost/healthz || exit 1
```

Now i have this configuration:

![Bildschirmfoto 2024-04-26 um 11 01 05](https://github.com/coollabsio/coolify/assets/20899975/9cf4d73b-2eb7-4fa0-b866-e0746c53b693)

The two problems i have are:
- It looks like the provided route is not called. When i look into the logs, i don't see any logs about that route. When i had the healthcheck in my dockerfile i could the the logs of the healthz route being called every 10 seconds
- I'm unsure about the `start period`, because in my deployment logs the health check is performed immediately and the attempts are counted up. Is this value used?

In my health check i check the database connection. As a test i shut down the database. 5 minutes later the backend resource was still marked as "healthy" so i think the health check is not really executing 

### Minimal Reproduction (if possible, example repository)

Please let me know if you need a reproduction and what this should contain

### Version

v4.0.0-beta.266

**Comment by pagoru** [comment]
Same problem on 270

**Comment by EinfachHans** [comment]
Will to re-test after the release and let you know if it works, thanks

**Comment by andrasbacsai** [comment]
Let me know if it still does not work

**Comment by andresledo** [comment]
> Let me know if it still does not work

Hello Andras, 

In the latest version v4.0.0-beta.271 all healtcheck give me error. If I disable the healcheck and do a curl to the URL from inside the container it responds me correctly. 

I would say that something is not working in this latest version. 

```
[2024-Apr-30 14:42:46.736785] Healthcheck URL (inside the container): GET: [http://localhost:80/](http://localhost/)
[2024-Apr-30 14:42:46.739129] Waiting for the start period (5 seconds) before starting healthcheck.
[2024-Apr-30 14:42:51.961738]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:42:51.968983] Attempt 1 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:42:57.166076]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:42:57.182480] Attempt 2 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:02.386444]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:02.409387] Attempt 3 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:07.594036]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:07.602918] Attempt 4 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:12.793358]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:12.815862] Attempt 5 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:17.970307]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:17.990975] Attempt 6 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:23.170582]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:23.185569] Attempt 7 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:28.418971]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:28.427266] Attempt 8 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:33.652387]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:33.658096] Attempt 9 of 10 | Healthcheck status: "starting"
[2024-Apr-30 14:43:38.834104]

[COMMAND] docker inspect --format='{{json .State.Health.Status}}' nsgkgkw-144216520989
[OUTPUT]
"starting"

[2024-Apr-30 14:43:38.845367] Attempt 10 of 10 | Healthcheck status: "starting"


```

![Captura de pantalla 2024-04-30 165003](https://github.com/coollabsio/coolify/assets/44968815/2f5e88b4-3a09-4872-8aca-3e089f4cd29b)

**Comment by fcpauldiaz** [comment]
I tried to re deploy my app today, without code changes and the health check is having error. I'm on v4.0.0-beta.271

**Comment by fcpauldiaz** [comment]
I added a log when accessing the root path, that should be the default health path.

This is how it looks a deployment, first is the current docker and the second one is the one being deployed and the print "GET /", doesn't show on the second one, meaning it doesn't call the health url.



![Screenshot 2024-04-30 at 6 40 37 PM](https://github.com/coollabsio/coolify/assets/4523284/3c6525e1-149f-4bb5-b496-89237643af89)

**Comment by andrasbacsai** [comment]
Can you please show me the healthcheck configuration page?

Also, is your app based on dockerfile buildpack?

**Comment by EinfachHans** [comment]
Hey @andrasbacsai ,

i also checked my health check now with 271. I use the same health check configuration as in my original post:

![Bildschirmfoto 2024-04-26 um 11 01 05](https://github.com/coollabsio/coolify/assets/20899975/9cf4d73b-2eb7-4fa0-b866-e0746c53b693)

My issues seems to be resolved now. The startup time is now awaited and i now see the logs of my defined route, which is called in the interval i provided! 👍🏼

I have the following points still in mind:
1. Instead of awaiting the startup time, you could already start checking for the health, but in case it fails, don't count up the attempt. Like this it could be marked as `healthy` quicker with still having a startup time where its "okay" for the healthcheck to fail
2. In my health check i do only a simple check if my db is accessible atm. When i turn off the db, the following happens: I stop seeing the logs about the request to the health url, **but** it seems to still call them, because the container is marked as unhealthy after a time. It also gets healthy after i turn on the db again. This is awesome, only thing missing here would be notifications for the health status change

**Comment by fcpauldiaz** [comment]
My app is based on Dockerfile

![Screenshot 2024-05-01 at 8 50 13 AM](https://github.com/coollabsio/coolify/assets/4523284/dea5c71d-be68-4614-
9792-4f4407145962)

**Comment by pagoru** [comment]
Still not working on me neither. (I disabled it because it fails)

![image](https://github.com/coollabsio/coolify/assets/1551265/34ebea8e-3199-4f80-9524-ca798e46462a)

**Comment by gkibria** [comment]
Hi, why I am not getting any healthchecks option in the ui? I tried docker compose project. Any idea? using 272. Thanks

**Comment by EinfachHans** [comment]
I just set up a new project where the health check is configured the same as in the other and it does not work..

**Comment by andrasbacsai** [comment]
No matter how I try to replicate this issue, I cannot.

Healthcheck is working on older version of Coolify version with Dockerfile buildpack, because there was basically no healthcheck (exit 0 command was used, so it always returns true).

@lewisdewsbury Your nextjs applications returns a 200 on path / ?

**Comment by EinfachHans** [comment]
But could you reproduce my two problems?

1. Missing notification in case of healthy status changes
2. Somehow missing logs for failed health checks

Also i would like to hear your opinion about my suggestion to the startPeriod to not await it, but don't count up the attempts while in this period.

**Comment by thomasmol** [comment]
I had the exact same problem as @lewisdewsbury. Started having the healthcheck not working on version `beta.271` on a resource that is build with a dockerfile. Dockerfile is based on a Bun base image. Added `apt-get install -y curl` in the dockerfile and now the healthcheck runs fine again! thanks @lewisdewsbury for the solution

**Comment by fcpauldiaz** [comment]
I had the same issue. @andrasbacsai you can close the ticket now.

**Comment by andrasbacsai** [comment]
I will make a few changes. Coolify will check if curl or wget is available. If neither, it will try to use the proc filesystem for hc, or just pass healthy state.

**Comment by EinfachHans** [comment]
> But could you reproduce my two problems?
> 
> 1. Missing notification in case of healthy status changes
> 2. Somehow missing logs for failed health checks
> 
> Also i would like to hear your opinion about my suggestion to the startPeriod to not await it, but don't count up the attempts while in this period.

@andrasbacsai did you already looked at these three points?

**Comment by EinfachHans** [comment]
@andrasbacsai 

**About 1**: Yes, this notification would be awesome! I would be notified directly if my service is unhealthy and i can the look into it why.

**About 2**: My application is an express backend. I log every request. Successfully requests about the health url are logged correctly. When i turn off my postgres, which causes the health check to return an error, the container status is updated correctly (so the requests are send i guess), but there are absolut no logs about these failed requests in my logs. When i then manually call the health check url via postman there is a log. Thats quite weird

**About 3**: Thats my suggestion to faster have a healthy container. By not awaiting the start period, but not count up failed attempts this could be arcvhieved.

**Comment by EinfachHans** [comment]
@andrasbacsai should i create a separate issue for these points?

**Comment by EinfachHans** [comment]
I opened a feature discussion about point 1 & 3: https://github.com/coollabsio/coolify/discussions/2200

**Comment by vikyw89** [comment]
On v4.0.0-beta.276

![image](https://github.com/coollabsio/coolify/assets/112059651/6ebd68dd-41ce-4e4b-9a7f-40bcb617b2b1)

**Comment by vikyw89** [comment]
> dockerfile. Dockerfile is based on a Bun base image. Added `apt-get install -y curl` in the dockerfile and now the healthcheck runs fine again! thanks @lewisdewsbury for the solution

so do we need to do the healthcheck from inside the container ? I always thought healthcheck should be observed from outside the docker container.

**Comment by vikyw89** [comment]
> No matter how I try to replicate this issue, I cannot.
> 
> Healthcheck is working on older version of Coolify version with Dockerfile buildpack, because there was basically no healthcheck (exit 0 command was used, so it always returns true).
> 
> @lewisdewsbury Your nextjs applications returns a 200 on path / ?

Is there a way to get the old "healthcheck" ?

**Comment by fcpauldiaz** [comment]
previously it wasn't doing anything, you can disable the healthcheck

**Comment by vikyw89** [comment]
> previously it wasn't doing anything, you can disable the healthcheck

before:
- i can deploy and have the old application replaced after the new one is ready, no downtime

currently:
- disabled healthcheck, application experiences downtime while the new app is starting. And in case the new deployment fail, the app is down.

**Comment by Jeyanthinath** [comment]
@andrasbacsai trying to understand, is the healthcheck is triggered from within the container which I built ? do I need to add curl/wget package in my final stage of app container build ? 

if so why its trying to execute from inside ?

**Comment by steveadams** [comment]
> > dockerfile. Dockerfile is based on a Bun base image. Added `apt-get install -y curl` in the dockerfile and now the healthcheck runs fine again! thanks [@lewisdewsbury](https://github.com/lewisdewsbury) for the solution
> 
> so do we need to do the healthcheck from inside the container ? I always thought healthcheck should be observed from outside the docker container.

Yes, I would expect the health check to be performed by the server that the resource belongs to.

**Comment by peaklabs-dev** [comment]
@steveadams Docker health checks run on the container, not the server, so if you have a curl health check, you must have curl installed in the container.
